### PR TITLE
Performance Improvements

### LINKED ISSUE #1397: dramatic performance degradation between JSqlParser 4.0 and 4.2
Some of our included PRs had a dramatic performance impact. So they need to be revisited. This log comes from the same SQL which uses non recursive large case when statements.

JSqlParser 4.2

```log
466091 - execution(private boolean net.sf.jsqlparser.parser.CCJSqlParser.jj_3R_SpecialStringFunctionWithNamedParameters_4486_9_249())
869250 - execution(private boolean net.sf.jsqlparser.parser.CCJSqlParser.jj_3R_Column_1597_5_134())
870821 - execution(private boolean net.sf.jsqlparser.parser.CCJSqlParser.jj_3R_UserVariable_3831_5_215())
870821 - execution(private boolean net.sf.jsqlparser.parser.CCJSqlParser.jj_3R_UserVariable_3831_12_456())
1335347 - execution(private boolean net.sf.jsqlparser.parser.CCJSqlParser.jj_3R_RelObjectNameList_1584_5_323())
1587076 - execution(private boolean net.sf.jsqlparser.parser.CCJSqlParser.jj_3R_RelObjectName_1656_5_174())
1587076 - execution(private boolean net.sf.jsqlparser.parser.CCJSqlParser.jj_3R_RelObjectNameExt_1686_7_409())
1587076 - execution(private boolean net.sf.jsqlparser.parser.CCJSqlParser.jj_3R_RelObjectName_1656_6_396())
1587076 - execution(private boolean net.sf.jsqlparser.parser.CCJSqlParser.jj_3R_RelObjectNameExt_1686_5_189())
1587076 - execution(private boolean net.sf.jsqlparser.parser.CCJSqlParser.jj_3R_RelObjectNameWithoutValue_1612_5_601())
177987285 - execution(private boolean net.sf.jsqlparser.parser.CCJSqlParser.jj_scan_token(int))
```

JSqlParser 4.0

```log
75789 - execution(private boolean net.sf.jsqlparser.parser.CCJSqlParser.jj_3_28())
75995 - execution(private boolean net.sf.jsqlparser.parser.CCJSqlParser.jj_3R_Column_1307_5_114())
76148 - execution(private boolean net.sf.jsqlparser.parser.CCJSqlParser.jj_3R_UserVariable_3334_5_175())
76148 - execution(private boolean net.sf.jsqlparser.parser.CCJSqlParser.jj_3R_UserVariable_3334_12_358())
116525 - execution(private boolean net.sf.jsqlparser.parser.CCJSqlParser.jj_3R_RelObjectNameList_1294_5_226())
154419 - execution(private boolean net.sf.jsqlparser.parser.CCJSqlParser.jj_3R_RelObjectName_1350_5_133())
154419 - execution(private boolean net.sf.jsqlparser.parser.CCJSqlParser.jj_3R_RelObjectName_1350_6_295())
154419 - execution(private boolean net.sf.jsqlparser.parser.CCJSqlParser.jj_3R_RelObjectNameExt_1380_5_147())
154419 - execution(private boolean net.sf.jsqlparser.parser.CCJSqlParser.jj_3R_RelObjectNameWithoutValue_1322_5_415())
154419 - execution(private boolean net.sf.jsqlparser.parser.CCJSqlParser.jj_3R_RelObjectNameExt_1380_7_308())
9681645 - execution(private boolean net.sf.jsqlparser.parser.CCJSqlParser.jj_scan_token(int))
```

If you look at the methods call log, some methods are called 10 times and more of the version 4.0.

Maybe we have to rewind some PRs for this.

### LINKED ISSUE #1438: Complex Expression Performance Issue reported by DBeaver
```sql
SELECT 	* FROM TABLE_1 t1
WHERE
	(((t1.COL1 = 'VALUE2' )
		AND (t1.CAL2 = 'VALUE2' ))
		AND (((1 = 1 )
			AND ((((((t1.id IN (940550 ,940600 ,940650 ,940700 ,940750 ,940800 ,940850 ,940900 ,940950 ,941000 ,941050 ,941100 ,941150 ,941200 ,941250 ,941300 ,941350 ,941400 ,941450 ,941500 ,941550 ,941600 ,941650 ,941700 ,941750 ,941800 ,941850 ,941900 ,941950 ,942000 ,942050 ,942100 ,942150 ,942200 ,942250 ,942300 ,942350 ,942400 ,942450 ,942500 ,942550 ,942600 ,942650 ,942700 ,942750 ,942800 ,942850 ,942900 ,942950 ,943000 ,943050 ,943100 ,943150 ,943200 ,943250 ,943300 ,943350 ,943400 ,943450 ,943500 ,943550 ,943600 ,943650 ,943700 ,943750 ,943800 ,943850 ,943900 ,943950 ,944000 ,944050 ,944100 ,944150 ,944200 ,944250 ,944300 ,944350 ,944400 ,944450 ,944500 ,944550 ,944600 ,944650 ,944700 ,944750 ,944800 ,944850 ,944900 ,944950 ,945000 ,945050 ,945100 ,945150 ,945200 ,945250 ,945300 ))
				OR (t1.id IN (945350 ,945400 ,945450 ,945500 ,945550 ,945600 ,945650 ,945700 ,945750 ,945800 ,945850 ,945900 ,945950 ,946000 ,946050 ,946100 ,946150 ,946200 ,946250 ,946300 ,946350 ,946400 ,946450 ,946500 ,946550 ,946600 ,946650 ,946700 ,946750 ,946800 ,946850 ,946900 ,946950 ,947000 ,947050 ,947100 ,947150 ,947200 ,947250 ,947300 ,947350 ,947400 ,947450 ,947500 ,947550 ,947600 ,947650 ,947700 ,947750 ,947800 ,947850 ,947900 ,947950 ,948000 ,948050 ,948100 ,948150 ,948200 ,948250 ,948300 ,948350 ,948400 ,948450 ,948500 ,948550 ,948600 ,948650 ,948700 ,948750 ,948800 ,948850 ,948900 ,948950 ,949000 ,949050 ,949100 ,949150 ,949200 ,949250 ,949300 ,949350 ,949400 ,949450 ,949500 ,949550 ,949600 ,949650 ,949700 ,949750 ,949800 ,949850 ,949900 ,949950 ,950000 ,950050 ,950100 )))
				OR (t1.id IN (950150 ,950200 ,950250 ,950300 ,950350 ,950400 ,950450 ,950500 ,950550 ,950600 ,950650 ,950700 ,950750 ,950800 ,950850 ,950900 ,950950 ,951000 ,951050 ,951100 ,951150 ,951200 ,951250 ,951300 ,951350 ,951400 ,951450 ,951500 ,951550 ,951600 ,951650 ,951700 ,951750 ,951800 ,951850 ,951900 ,951950 ,952000 ,952050 ,952100 ,952150 ,952200 ,952250 ,952300 ,952350 ,952400 ,952450 ,952500 ,952550 ,952600 ,952650 ,952700 ,952750 ,952800 ,952850 ,952900 ,952950 ,953000 ,953050 ,953100 ,953150 ,953200 ,953250 ,953300 ,953350 ,953400 ,953450 ,953500 ,953550 ,953600 ,953650 ,953700 )))
				OR (t1.id IN (953750 ,953800 ,953850 ,953900 ,953950 ,954000 ,954050 ,954100 ,954150 ,954200 ,954250 ,954300 ,954350 ,954400 ,954450 ,954500 ,954550 ,954600 ,954650 ,954700 ,954750 ,954800 ,954850 ,954900 ,954950 ,955000 ,955050 ,955100 ,955150 ,955200 ,955250 ,955300 ,955350 ,955400 ,955450 ,955500 ,955550 ,955600 ,955650 ,955700 ,955750 ,955800 ,955850 ,955900 ,955950 ,956000 ,956050 ,956100 ,956150 ,956200 ,956250 ,956300 ,956350 ,956400 ,956450 ,956500 ,956550 ,956600 ,956650 ,956700 ,956750 ,956800 ,956850 ,956900 ,956950 ,957000 ,957050 ,957100 ,957150 ,957200 ,957250 ,957300 )))
				OR (t1.id IN (944100, 944150, 944200, 944250, 944300, 944350, 944400, 944450, 944500, 944550, 944600, 944650, 944700, 944750, 944800, 944850, 944900, 944950, 945000 )))
				OR (t1.id IN (957350 ,957400 ,957450 ,957500 ,957550 ,957600 ,957650 ,957700 ,957750 ,957800 ,957850 ,957900 ,957950 ,958000 ,958050 ,958100 ,958150 ,958200 ,958250 ,958300 ,958350 ,958400 ,958450 ,958500 ,958550 ,958600 ,958650 ,958700 ,958750 ,958800 ,958850 ,958900 ,958950 ,959000 ,959050 ,959100 ,959150 ,959200 ,959250 ,959300 ,959350 ,959400 ,959450 ,959500 ,959550 ,959600 ,959650 ,959700 ,959750 ,959800 ,959850 ,959900 ,959950 ,960000 ,960050 ,960100 ,960150 ,960200 ,960250 ,960300 ,960350 ,960400 ,960450 ,960500 ,960550 ,960600 ,960650 ,960700 ,960750 ,960800 ,960850 ,960900 ,960950 ,961000 ,961050 ,961100 ,961150 ,961200 ,961250 ,961300 ,961350 ,961400 ,961450 ,961500 ,961550 ,961600 ,961650 ,961700 ,961750 ,961800 ,961850 ,961900 ,961950 ,962000 ,962050 ,962100 ))))
				OR (t1.id IN (962150 ,962200 ,962250 ,962300 ,962350 ,962400 ,962450 ,962500 ,962550 ,962600 ,962650 ,962700 ,962750 ,962800 ,962850 ,962900 ,962950 ,963000 ,963050 ,963100 ,963150 ,963200 ,963250 ,963300 ,963350 ,963400 ,963450 ,963500 ,963550 ,963600 ,963650 ,963700 ,963750 ,963800 ,963850 ,963900 ,963950 ,964000 ,964050 ,964100 ,964150 ,964200 ,964250 ,964300 ,964350 ,964400 ,964450 ,964500 ,964550 ,964600 ,964650 ,964700 ,964750 ,964800 ,964850 ,964900 ,964950 ,965000 ,965050 ,965100 ,965150 ,965200 ,965250 ,965300 ,965350 ,965400 ,965450 ,965500 ))))
	AND t1.COL3 IN (
	    SELECT
		    t2.COL3
	    FROM
		    TABLE_6 t6,
		    TABLE_1 t5,
		    TABLE_4 t4,
		    TABLE_3 t3,
		    TABLE_1 t2
	    WHERE
		    (((((((t5.CAL3 = T6.id)
			    AND (t5.CAL5 = t6.CAL5))
			    AND (t5.CAL1 = t6.CAL1))
			    AND (t3.CAL1 IN (108500)))
			    AND (t5.id = t2.id))
			    AND NOT ((t6.CAL6 IN ('VALUE'))))
			    AND ((t2.id = t3.CAL2)
				    AND (t4.id = t3.CAL3))))
ORDER BY
	t1.id ASC
```

### PR DESCRIPTION
**Pull request by manticore-projects**
Fixes #1397
Fixes #1438
Try to parse simple and fast first, before parsing complex
Update libraries to its latest versions
Remove JUnit4 Leftover

Add a 6 second timeout to the Parser and let it fail gracefully when a Statement can't be parsed within that time
Add a special test assuring the cleanup of the Parser objects after the timeout (no stale background processes)

### CODE REVIEW

File: src/main/java/net/sf/jsqlparser/parser/CCJSqlParserUtil.java [hunk 1]
```diff
@@ -54,13 +63,25 @@ public static Statement parse(String sql) throws JSQLParserException {
...
-        boolean allowComplexParsing = getNestingDepth(sql)<=ALLOWED_NESTING_DEPTH;
-        
-        CCJSqlParser parser = newParser(sql).withAllowComplexParsing(allowComplexParsing);
-        if (consumer != null) {
-            consumer.accept(parser);
+        Statement statement = null;
+
+        // first, try to parse fast and simple
+        try {
+            CCJSqlParser parser = newParser(sql).withAllowComplexParsing(false);
```
Comment (wumpz) [inline src/main/java/net/sf/jsqlparser/parser/CCJSqlParserUtil.java:70]: If complex parsing is initially switched off, it should not be switched on in this method. Right? However, this is only some kind of hack. The grammar itself is flawed if we need to do something like this. The porblem is the excessive use of Lookups, but if the grammar would be designed that a fixed lookup would be enough, then we do not need a switch like this.

**Review by wumpz — COMMENTED** [review]
I think the problem lies deeper. So this PR does somehow covers the symptoms but not the reasons. So we need to dig deeper.

File: build.gradle [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
+    id "ca.coglinc2.javacc" version "latest.release"
     id 'jacoco'
-    id "com.github.spotbugs" version "4.7.2"
+    id "com.github.spotbugs" version "latest.release"
     id 'pmd'
     id 'checkstyle'
     
     // download the RR tools which have no Maven Repository
-    id "de.undercouch.download" version "4.1.2"
+    id "de.undercouch.download" version "latest.release"
```
[the remark this answers is not in the document]
Comment (manticore-projects) [reply build.gradle:11]: I do not know about Maven does (In fact I hate Maven. You can't even `clean` a project while offline. I went straight from ANT to Gradle, which "just understands and does what I want" instead of forcing me into any dogma). This `latest.release` automatically retrieves the Latest Release of a Gradle Plugin.

File: build.gradle [hunk 3]
```diff
@@ -23,29 +23,29 @@ repositories {
...
 
 dependencies {
-    testImplementation 'commons-io:commons-io:2.6'
-    testImplementation 'junit:junit:4.13.1'
-    testImplementation 'org.mockito:mockito-core:2.28.2'
-    testImplementation 'org.assertj:assertj-core:3.16.1'
-    testImplementation 'org.apache.commons:commons-lang3:3.10'
-    testImplementation 'com.h2database:h2:1.4.200'
+    testImplementation 'commons-io:commons-io:2.11.0+'
+    testImplementation 'org.mockito:mockito-core:4.1.0+'
```
[the remark this answers is not in the document]
Comment (manticore-projects) [reply build.gradle:34]: Again, this is Gradle Sugar, which allows to use the latest update of a Library following "semantic versioning". In fact I applied it wrong. It should be either `testImplementation 'commons-io:commons-io:2.11.+' for fetching latest release of the `2.11` minor version or alternatively  `testImplementation 'commons-io:commons-io:2.+' for fetching the latest minor version of the `2` major version.

Found that also very handy for quickly updating Log4J2 recently :)

File: src/main/java/net/sf/jsqlparser/parser/CCJSqlParserUtil.java (hunk 1 above)
Comment (manticore-projects) [reply src/main/java/net/sf/jsqlparser/parser/CCJSqlParserUtil.java:70]: You are correct in your observation, but I respectfully do not share your conclusion:

1) It is not a `hack` but actually a completely valid approach (to my knowledge the only available method) to implement a `IF ... THEN ...` condition in the parser

2) While the grammar certainly could be rewritten/optimized, I still claim that SQL is a bit exceptional:
a) we try to be RDBMS/Dialect agnostic and so have to deal with a great flexibility in the Grammar
b) SQL lacks reliable `terminated Blocks`, which makes it difficult to define `Stop` tokens. What do I mean with this: Obviously everything inside brackets `(` `)` is a block and the closing brackets `)` terminates this block and we know that we can stop parsing an expression at the closing bracket. Unfortunately SQL was "designed" for assistant secretaries, who were supposed to query data without programming skills.

Best illustration is issue #1444:
```sql
select doc.fullName from XWikiDocument doc where doc.fullName like :fullName escape '/' and doc.author like :author escape '!'
```

Is this
```sql
select doc.fullName from XWikiDocument doc where ( doc.fullName like :fullName escape '/' ) 
and ( doc.author like :author escape '!' )
```

Or is this
```sql
select doc.fullName from XWikiDocument doc where doc.fullName like :fullName escape ( '/' 
and doc.author like :author escape '!' )
```

JavaCC by its nature lacks a lot of functionality to deal with those requirements. JavaCC 21 added some features for a reason and I believe ANTLR is there for a reason too.

Long story short:
Yes, you are right: too many `lookup` harms us badly. But I am also right: we do need those  for parsing certain valid statements (look how functionality has grown since 4.0 and how the "new issue" rate went down since!)

So the suggest approach, while not a beauty at all solves that in a pragmatic way:
It tries to parse first without those slow `lookup` and only when this fails, it would activate the `lookup` -- because they are needed.

Comment (manticore-projects) [reply src/main/java/net/sf/jsqlparser/parser/CCJSqlParserUtil.java:70]: Of course, if anyone can point me on how to avoid those `lookup` without losing functionality, I am all in!

**Comment by manticore-projects** [comment]
Totally unrelated: We pull the complete ANTLR as dependency for PMD, while we work with JavaCC in order to not use ANTLR -- I find that quite funny :)

<pre><font color="#4E9A06">pmd</font><font color="#C4A000"> - The PMD libraries to be used for this project.</font>
<font color="#C4A000">\--- </font>net.sourceforge.pmd:pmd-java:6.41.0
<font color="#C4A000">     +--- </font>net.sourceforge.pmd:pmd-core:6.41.0
<font color="#C4A000">     |    +--- </font>org.antlr:antlr4-runtime:4.7.2
<font color="#C4A000">     |    +--- </font>com.beust:jcommander:1.48
<font color="#C4A000">     |    +--- </font>commons-io:commons-io:2.6
<font color="#C4A000">     |    +--- </font>net.sourceforge.saxon:saxon:9.1.0.8
<font color="#C4A000">     |    +--- </font>org.apache.commons:commons-lang3:3.8.1
<font color="#C4A000">     |    +--- </font>org.ow2.asm:asm:9.2
<font color="#C4A000">     |    \--- </font>com.google.code.gson:gson:2.8.5
<font color="#C4A000">     +--- </font>net.sourceforge.saxon:saxon:9.1.0.8
<font color="#C4A000">     +--- </font>org.ow2.asm:asm:9.2
<font color="#C4A000">     +--- </font>commons-io:commons-io:2.6
<font color="#C4A000">     \--- </font>org.apache.commons:commons-lang3:3.8.1
</pre>

File: src/test/resources/net/sf/jsqlparser/statement/select/oracle-tests/cast_multiset13.sql [hunk 4]
```diff
@@ -11,4 +11,5 @@ select customer_id, cust_address_ntab
 multiset except distinct cust_address2_ntab multiset_except
 from customers_demo
 
---@FAILURE: Encountered unexpected token: "except" "EXCEPT" recorded first on Aug 3, 2021, 7:20:08 AM
\ No newline at end of file
+--@FAILURE: Encountered unexpected token: "except" "EXCEPT" recorded first on Aug 3, 2021, 7:20:08 AM
+--@FAILURE: net.sf.jsqlparser.parser.ParseException: Encountered unexpected token: "except" "EXCEPT" recorded first on 21 Dec 2021, 15:15:16
```
Comment (wumpz) [inline src/test/resources/net/sf/jsqlparser/statement/select/oracle-tests/cast_multiset13.sql:15]: Why did you change this failure lines?

**Comment by manticore-projects** [comment]
Sorry, I am lost -- too many branches and PRs open for too long time.
I have reverted those unwarranted changes already in my local repo and pushed it again, but somehow it is not updated in the PR.

**Comment by VanillaIceCreams** [comment]
Why create a single-threaded thread pool each time the parseStatement method is called? And also immediately synchronous blocking! How is this different from executing directly in the current thread?
 When 200 concurrent requests came into one of my paginated query interfaces, the number of threads I served directly doubled ! wtf !

**Comment by VanillaIceCreams** [comment]
> Why create a single-threaded thread pool each time the parseStatement method is called? And also immediately synchronous blocking! How is this different from executing directly in the current thread? When 200 concurrent requests came into one of my paginated query interfaces, the number of threads I served directly doubled ! wtf !

Is the purpose to throw an exception after an 8-second timeout? Why is the performance of this method so bad that it can not be executed in 8 seconds?
